### PR TITLE
Implement support for the `[Exposed]` extended attribute

### PR DESCRIPTION
**Pull request by ExE-Boss**
Extracted from <https://github.com/jsdom/webidl2js/pull/190> at @TimothyGu’s request in <https://github.com/jsdom/webidl2js/pull/191#pullrequestreview-387619767>.

This also fixes a bug where `install` takes `globalName` instead of `globalNames`.

---

This will free up <https://github.com/jsdom/webidl2js/pull/190> to only handle the creation of `webidl2js‑globals.js`.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -454,6 +454,7 @@ webidl2js is implementing an ever-growing subset of the Web IDL specification. S
 - Variadic arguments
 - `[Clamp]`
 - `[EnforceRange]`
+- `[Exposed]`
```
Comment (TimothyGu) [inline README.md:457]: Make sure to remove this from the list of unsupported features.

File: lib/constructs/callback-interface.js [hunk 2]
```diff
@@ -178,14 +197,41 @@ class CallbackInterface {
...
+        const exposed = new Set([
+      `;
+
+      for (const globalName of this.exposed) {
+        this.str += `"${globalName}",\n`;
+      }
+
+      this.str += `
+        ]);
+      `;
```
Comment (TimothyGu) [inline lib/constructs/callback-interface.js:211]: What about just `const exposed = new Set(${JSON.stringify([...this.exposed])});`

File: lib/constructs/callback-interface.js [hunk 3]
```diff
@@ -178,14 +197,41 @@ class CallbackInterface {
...
+        for (const globalName of globalNames) {
+          if (exposed.has(globalName)) {
+            isExposed = true;
+            break;
+          }
+        }
+
+        if (!isExposed) {
+          return;
+        }
```
Comment (TimothyGu) [inline lib/constructs/callback-interface.js:233]: Can we simplify it to just `if (!globalNames.some(name => exposed.has(name)) return;`

File: lib/constructs/interface.js [hunk 4]
```diff
@@ -1463,7 +1463,47 @@ class Interface {
...
-      exports.install = (globalObject, globalName) => {
+      const exposed = new Set([
+    `;
+
+    for (const globalName of this.exposed) {
+      this.str += `"${globalName}",\n`;
+    }
+
+    this.str += `
+      ]);
```
Comment (TimothyGu) [inline lib/constructs/interface.js:1474]: Same comments as above.

File: lib/constructs/interface.js [hunk 5]
```diff
@@ -1463,7 +1463,47 @@ class Interface {
...
+        for (const globalName of globalNames) {
+          if (exposed.has(globalName)) {
+            isExposed = true;
+    `;
+
+    if (this.legacyWindowAliases) {
+      this.str += `
+            if (globalName === "Window") {
+              isWindow = true;
+            }
```
Comment (TimothyGu) [inline lib/constructs/interface.js:1494]: Same comments as above. For the `isWindow` check, another `globalNames.some()` should be fine too.

File: lib/constructs/callback-interface.js (hunk 3 above)
Comment (ExE-Boss) [reply lib/constructs/callback-interface.js:233]: `globalNames` is a `Set` in <https://github.com/jsdom/webidl2js/pull/190>.

Comment (TimothyGu) [reply lib/constructs/callback-interface.js:233]: Ah, someone should really make a `Set.prototype.some`/`every` proposal to TC39. But I'd rather the API be taking an array.

Comment (ExE-Boss) [reply lib/constructs/callback-interface.js:225]: There is the **Iterator Helpers** proposal: <https://tc39.es/proposal-iterator-helpers/>.

**Pull request merged**